tests/plugins/test_plugin.py

```
@@ -0,0 +1,127 @@
+
+
+@mock.patch.dict(
+    os.environ,
+    {
+        "CUDA_VISIBLE_DEVICES": "0,1",
+        "SLURM_NTASKS": "2",
+        "SLURM_JOB_NAME": "SOME_NAME",
+        "SLURM_NODEID": "0",
+        "LOCAL_RANK": "0",
+        "SLURM_LOCALID": "0",
+    },
+)
+@mock.patch("torch.cuda.device_count", return_value=2)
+@pytest.mark.parametrize(
+    ["ddp_backend", "gpus", "num_processes"],
+    [("ddp_cpu", None, None), ("ddp", 2, 0), ("ddp2", 2, 0), ("ddp_spawn", 2, 0)],
+)
+def test_custom_required_plugins(tmpdir, ddp_backend, gpus, num_processes):
+    """
+        Test to ensure that if a plugin requires certain plugin to be added, these are added automatically
+    """
+
+    class RequiredPlugin(NativeAMPPlugin):
+        """
+            My custom amp plugin that's required with my DDP plugin as default.
+            This allows us to ensure this plugin is added when using CustomPlugin rather than ensuring
+            the user passes it manually into the list.
+        """
+
+    class CustomPlugin(DDPPlugin):
+        def required_plugins(self, amp_backend: AMPType) -> Optional[list]:
+            return [RequiredPlugin()]
+
+    class CB(Callback):
+        def on_fit_start(self, trainer, pl_module):
+            assert isinstance(trainer.accelerator_backend.ddp_plugin, CustomPlugin)
+            assert isinstance(trainer.precision_connector.backend, RequiredPlugin)
+
+            raise SystemExit()
+
+    model = BoringModel()
+    with pytest.warns(UserWarning,
+                      match=f'plugin {type(CustomPlugin())} has added additional '
+                            f'required plugins as default: {[type(RequiredPlugin())]}*'):
+        trainer = Trainer(
+            fast_dev_run=True,
+            gpus=gpus,
+            num_processes=num_processes,
+            distributed_backend=ddp_backend,
+            plugins=[CustomPlugin()],
+            callbacks=[CB()],
+        )
+    with pytest.raises(SystemExit):
+        trainer.fit(model)
```

not sure about this one, what is the meaning, shall we rather raise an exception instead exit?